### PR TITLE
Enforce wildcard singleton in superuser whitelist

### PR DESCRIPTION
**Pull request by mpalmi**

### CODE REVIEW

File: set_user.c [hunk 1]
```diff
@@ -318,12 +318,15 @@ set_user(PG_FUNCTION_ARGS)
 				/* can only escalate with set_user_u */
 				ereport(ERROR,
 						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-						 errmsg("Switching to superuser only allowed for privileged procedure: \'set_user_u\'")));
+						 errmsg("Switching to superuser only allowed "
```
Comment (jconway) [inline set_user.c]: Per the error message guidelines, this message should be more like "switching to superuser not allowed" with perhaps an errhint suggesting "use set_user_u() instead". Note that errmsg should not start with a capitalized word, nor be a complete sentence, also per the guidelines. See https://www.postgresql.org/docs/current/static/error-style-guide.html

File: set_user.c [hunk 2]
```diff
@@ -318,12 +318,15 @@ set_user(PG_FUNCTION_ARGS)
...
+								"for privileged procedure: \'set_user_u\'")));
 			else if (!check_user_whitelist(GetUserId(), SU_Whitelist))
 				/* check superuser whitelist*/
 				ereport(ERROR,
 						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-						 errmsg("Switching to superuser not allowed for current user")));
+						 errmsg("Switching to superuser not allowed for current user"),
```
Comment (jconway) [inline set_user.c]: Similar comment here -- no capital letter S

File: set_user.c [hunk 3]
```diff
@@ -182,18 +183,38 @@ check_user_whitelist(Oid userId, const char *whitelist)
...
+	/* Allow all users to escalate if whitelist is a solo wildcard character. */
+	if (list_length(elemlist) == 1)
+	{
+		char	   *first_elem = NULL;
+
+		first_elem = (char *) linitial(elemlist);
+		if (pg_strcasecmp(first_elem, WHITELIST_WILDCARD) == 0)
+			return true;
+	}
+
```
Comment (jconway) [inline set_user.c:195]: The entire next stanza should probably be inside"
else
{
    ....
}

File: set_user.c [hunk 4]
```diff
@@ -182,18 +183,38 @@ check_user_whitelist(Oid userId, const char *whitelist)
...
+						 errmsg("Invalid syntax in parameter"),
+						 errhint("Either remove users from set_user.superuser_whitelist "
+								 "or remove the wildcard character \"%s\". The whitelist "
+								 "cannot contain both.",
+								 WHITELIST_WILDCARD)));
 	}
 
-	return false;
+	return result;
 }
```
Comment (jconway) [inline set_user.c:219]: I think there are at least a couple places in the README that need subtle rewording based on this.

**Comment by mpalmi** [comment]
@jconway feedback addressed in https://github.com/pgaudit/set_user/pull/5/commits

**Pull request merged**